### PR TITLE
feat: reverse mode rules for fft

### PR DESCRIPTION
**Pull request by avik-pal**
- [x] implementation
- [x] verify numerical correctness in reactant -- rfft seems off
- [x] activate all the tests

based on https://github.com/jax-ml/jax/blob/ff6892b41541202ebcec2a791862dfc05910baca/jax/_src/lax/fft.py#L139

### CODE REVIEW

File: src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp [hunk 1]
```diff
@@ -217,10 +218,20 @@ template <typename T> Attribute makeAttr(mlir::Type elemType, T val) {
   if (auto TT = dyn_cast<RankedTensorType>(elemType))
     return SplatElementsAttr::get(
         TT, ArrayRef(makeAttr<T>(TT.getElementType(), val)));
-  if (isa<FloatType>(elemType))
+  if (isa<FloatType>(elemType)) {
     return FloatAttr::get(elemType, val);
```
Comment (wsmoses) [inline src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp]: this code is copied enough places that if it wasn't sufficeint everywhere else will also be broken. Can we move it to a header file to be used once everywhere (grep for makeAttr in the repo)

Comment (avik-pal) [reply src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp]: done

File: src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp [hunk 2]
```diff
@@ -3868,6 +3860,127 @@ struct StablehloSubSimplifyMathInterface
...
+class AutoDiffFftOpRev
+    : public ReverseAutoDiffOpInterface::ExternalModel<AutoDiffFftOpRev,
+                                                       FftOp> {
+public:
+  LogicalResult createReverseModeAdjoint(Operation *orig, OpBuilder &builder,
+                                         MGradientUtilsReverse *gutils,
+                                         SmallVector<Value> caches) const {
+    auto op = cast<stablehlo::FftOp>(orig);
+
+    if (gutils->isConstantValue(op->getResult(0)))
```
Comment (wsmoses) [inline src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp]: this should be automated presently before calling the fn

**Comment by avik-pal** [comment]
```mlir
// REVERSE-FFT:  func.func @fft(%arg0: tensor<4xcomplex<f32>>, %arg1: tensor<4xcomplex<f32>>) -> tensor<4xcomplex<f32>> {
// REVERSE-FFT-NEXT:    %cst = stablehlo.constant dense<(0.000000e+00,0.000000e+00)> : tensor<4xcomplex<f32>>
// REVERSE-FFT-NEXT:    %0 = stablehlo.add %cst, %arg1 : tensor<4xcomplex<f32>>
// REVERSE-FFT-NEXT:    %1 = chlo.conj %0 : tensor<4xcomplex<f32>> -> tensor<4xcomplex<f32>>
// REVERSE-FFT-NEXT:    %2 = stablehlo.fft %1, type =  FFT, length = [4] : (tensor<4xcomplex<f32>>) -> tensor<4xcomplex<f32>>
// REVERSE-FFT-NEXT:    %3 = chlo.conj %2 : tensor<4xcomplex<f32>> -> tensor<4xcomplex<f32>>
// REVERSE-FFT-NEXT:    %4 = stablehlo.add %cst, %3 : tensor<4xcomplex<f32>>
// REVERSE-FFT-NEXT:    return %4 : tensor<4xcomplex<f32>>
// REVERSE-FFT-NEXT:  }
```

where are the conj ops coming from? i don't see them in tablegen

**Comment by mofeing** [comment]
> writing in cpp, this seems a bit painful to write out in td 😓

yeah, i had the same impression when i tried.

do we support numerical tests on CI? there are a lot of edge cases due to the stablehlo.fft rules on batch dimensions and memory-saving for real fft/ifft

**Comment by mofeing** [comment]
> ```mlir
> // REVERSE-FFT:  func.func @fft(%arg0: tensor<4xcomplex<f32>>, %arg1: tensor<4xcomplex<f32>>) -> tensor<4xcomplex<f32>> {
> // REVERSE-FFT-NEXT:    %cst = stablehlo.constant dense<(0.000000e+00,0.000000e+00)> : tensor<4xcomplex<f32>>
> // REVERSE-FFT-NEXT:    %0 = stablehlo.add %cst, %arg1 : tensor<4xcomplex<f32>>
> // REVERSE-FFT-NEXT:    %1 = chlo.conj %0 : tensor<4xcomplex<f32>> -> tensor<4xcomplex<f32>>
> // REVERSE-FFT-NEXT:    %2 = stablehlo.fft %1, type =  FFT, length = [4] : (tensor<4xcomplex<f32>>) -> tensor<4xcomplex<f32>>
> // REVERSE-FFT-NEXT:    %3 = chlo.conj %2 : tensor<4xcomplex<f32>> -> tensor<4xcomplex<f32>>
> // REVERSE-FFT-NEXT:    %4 = stablehlo.add %cst, %3 : tensor<4xcomplex<f32>>
> // REVERSE-FFT-NEXT:    return %4 : tensor<4xcomplex<f32>>
> // REVERSE-FFT-NEXT:  }
> ```
> 
> where are the conj ops coming from? i don't see them in tablegen

i think they are automatically added on autodiff in order to support complex rev diff. AFAIK they are added always and then optimized aways, but in this case we have complex tensors so they can't be optimized away so easily.

**Comment by wsmoses** [comment]
Not a full explanation https://enzymead.github.io/Enzyme.jl/dev/faq/#Complex-numbers but essentially reverse mode propagates conjugates of derivatives. Tablegen will automatically add the conj to reverse mode rules

**Comment by avik-pal** [comment]
everything should now be in td

File: test/lit_tests/diffrules/stablehlo/fft.mlir [hunk 3]
```diff
@@ -51,8 +51,8 @@ func.func @ifft(%x : tensor<4xcomplex<f32>>) -> tensor<4xcomplex<f32>> {
 // REVERSE-IFFT-NEXT:    return %4 : tensor<4xcomplex<f32>>
 // REVERSE-IFFT-NEXT:  }
 
-// TODO enzymexlamlir-opt %s --enzyme-wrap="infn=rfft outfn= retTys=enzyme_dup argTys=enzyme_dup mode=ForwardMode" | FileCheck %s --check-prefix=FORWARD-RFFT
-// TODO enzymexlamlir-opt %s --enzyme-wrap="infn=rfft outfn= retTys=enzyme_active argTys=enzyme_active mode=ReverseModeCombined" --canonicalize --remove-unnecessary-enzyme-ops | FileCheck %s --check-prefix=REVERSE-RFFT
+// RUN: enzymexlamlir-opt %s --enzyme-wrap="infn=rfft outfn= retTys=enzyme_dup argTys=enzyme_dup mode=ForwardMode" | FileCheck %s --check-prefix=FORWARD-RFFT
+// TODO: enzymexlamlir-opt %s --enzyme-wrap="infn=rfft outfn= retTys=enzyme_active argTys=enzyme_active mode=ReverseModeCombined" --canonicalize --remove-unnecessary-enzyme-ops | FileCheck %s --check-prefix=REVERSE-RFFT
```
Comment (avik-pal) [inline test/lit_tests/diffrules/stablehlo/fft.mlir]: how do I run the postpasses here? otherwise it will fail:

```
test/lit_tests/diffrules/stablehlo/fft.mlir:58:8: error: 'complex.add' op operand #0 must be complex type with floating-point elements, but got 'tensor<3xcomplex<f32>>'
```

Comment (wsmoses) [reply test/lit_tests/diffrules/stablehlo/fft.mlir]: In principle just run the pass and perhaps disable a verify-after-all or similar flag.

alternatively rewrite it to call the regular enzyme pass rather than enzyme-wrap

**Comment by avik-pal** [comment]
numerically verified the results with the ChainRules using Reactant

**Review by wsmoses — APPROVED** [review]
